packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,87 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:collection';
+
+import 'asset_bundle.dart';
+
+/// Base class for resolver components that transform an input into part of
+/// a resolved asset's key.
+abstract class Qualifier {
+  String resolve();
+}
+
+/// Qualifier that knows how to find the nearest match among a list of
+/// candidates, based on a mapping to candidates from their normative values.
+/// The mapping should be sorted, e.g. by passing a [SplayTreeMap].
+///
+/// The resolution is done via linear search based on the assumption that the
+/// number of candidates is short.
```

Comment:
In general, linear search is quicker than binary search for fewer than N items where N depends on the language and implementation.  (Many production implementations of binary search use linear search below a threshold.)  Since as far as I know there's no built-in binary search operation in Dart, I'd either have to roll my own or import a separate package, either of which seemed like it was bordering on premature optimization.